joblib/parallel.py

```diff
@@ -86,23 +83,86 @@ def _register_dask():
 }
 
 
-def get_active_backend(prefer=None, require=None, verbose=0):
+# Sentinels for the default values of the Parallel constructor and
+# the parallel_config and parallel_backend context managers
+default_parallel_config = {
+    "n_jobs": _Sentinel(default_value=None),
+    "verbose": _Sentinel(default_value=0),
+    "temp_folder": _Sentinel(default_value=None),
+    "max_nbytes": _Sentinel(default_value="1M"),
+    "mmap_mode": _Sentinel(default_value="r"),
+    "prefer": _Sentinel(default_value=None),
+    "require": _Sentinel(default_value=None),
+}
+
+
+VALID_BACKEND_HINTS = ('processes', 'threads', None)
+VALID_BACKEND_CONSTRAINTS = ('sharedmem', None)
+
+
+def _get_config_param(param, context_config, key):
+    if param is not default_parallel_config[key]:
+        # param is explicitely set, return it
+        return param
+
+    if context_config is None or key not in context_config:
+        # no context manager just return the default value.
+        return param.default_value
+
+    if context_config[key] is not default_parallel_config[key]:
+        # there's a context manager and the key is set, return it
+        return context_config[key]
+
+    # there's a context manager but the key is not set, return the default
+    return param.default_value
+
+
+def get_active_backend(
+    prefer=default_parallel_config["prefer"],
```

Comment:
Currently, `_Sentinel` is a mutable object and we used it as a default parameter. It would be safer to make sure this is immutable. We can probably achieve this goal in several ways: using `__slots__` or decorating with the `dataclass` decorator (`frozen` could be useful there - maybe `slots` as well).

Reply:
used a frozen dataclass